src/axom/core/Array.hpp

```diff
@@ -1028,51 +1210,64 @@ inline typename Array<T>::ArrayIterator Array<T>::emplace(Array<T>::ArrayIterato
 //------------------------------------------------------------------------------
 template <typename T>
 template <typename... Args>
-inline void Array<T>::emplace_back(Args&&... args)
+inline void ArrayImpl<T, 1>::emplace_back(Args&&... args)
 {
-  emplace(m_num_elements, args...);
+  asDerived().emplace(asDerived().size(), args...);
 }
 
 //------------------------------------------------------------------------------
-template <typename T>
-inline void Array<T>::resize(IndexType new_num_elements)
+template <typename T, int DIM>
+template <typename... Args>
+inline void Array<T, DIM>::resize(Args... args)
 {
-  assert(new_num_elements >= 0);
+  static_assert(sizeof...(Args) == DIM,
+                "Array size must match number of dimensions");
+  assert(detail::allPositive(args...));
+  const auto new_num_elements = detail::packProduct(args...);
 
   if(new_num_elements > m_capacity)
   {
     dynamicRealloc(new_num_elements);
   }
 
   updateNumElements(new_num_elements);
+
+  static_cast<ArrayImpl<T, DIM>&>(*this) =
+    ArrayImpl<T, DIM> {static_cast<IndexType>(args)...};
 }
 
 //------------------------------------------------------------------------------
-template <typename T>
-inline void Array<T>::swap(Array<T>& other)
+template <typename T, int DIM>
+inline void Array<T, DIM>::swap(Array<T, DIM>& other)
 {
+  // FIXME: Why doesn't ADL work here?
+  // swap(static_cast<ArrayImpl<T, DIM>&>(*this),
+  //      static_cast<ArrayImpl<T, DIM>&>(other));
```

Comment:
I can never figure out how to do template friend functions. But template member functions are easy, and then you can make a free function `swap(Array x, Array y)` that just calls through to the member function `x.swap(y)`;

https://godbolt.org/z/aWqcar37e